packages/loopring_v3/contracts/aux/agents/FastWithdrawalAgent.sol

```diff
@@ -12,152 +12,180 @@ import "../../lib/ReentrancyGuard.sol";
+    function execute_(FastWithdrawalApproval memory fwa)
         internal
+        returns (Status)
     {
-        // The liquidity provider always authorizes the fast withdrawal by being the direct caller
-        address payable liquidityProvider = msg.sender;
-
-        if (fastWithdrawal.signature.length > 0) {
-            // Compute the hash
-            bytes32 hash = EIP712.hashPacked(
-                DOMAIN_SEPARATOR,
-                keccak256(
-                    abi.encodePacked(
-                        FASTWITHDRAWAL_TYPEHASH,
-                        fastWithdrawal.exchange,
-                        fastWithdrawal.from,
-                        fastWithdrawal.to,
-                        fastWithdrawal.token,
-                        fastWithdrawal.amount,
-                        fastWithdrawal.feeToken,
-                        fastWithdrawal.fee,
-                        fastWithdrawal.nonce,
-                        fastWithdrawal.validUntil
-                    )
+        require(
+            fwa.exchange != address(0) &&
+            fwa.from != address(0) &&
+            fwa.to != address(0) &&
+            fwa.amount != 0,
+            "INVALID_WITHDRAWAL"
+        );
+
+        // Compute the hash
+        bytes32 hash = EIP712.hashPacked(
+            DOMAIN_SEPARATOR,
+            keccak256(
+                abi.encodePacked(
+                    FASTWITHDRAWAL_APPROVAL_TYPEHASH,
+                    fwa.exchange,
+                    fwa.from,
+                    fwa.to,
+                    fwa.token,
+                    fwa.amount,
+                    fwa.storageID,
+                    fwa.validUntil
                 )
-            );
+            )
+        );
 
-            // Check the signature
-            require(hash.verifySignature(fastWithdrawal.from, fastWithdrawal.signature), "INVALID_SIGNATURE");
-
-            // Check the time limit
-            require(block.timestamp <= fastWithdrawal.validUntil, "TX_EXPIRED");
-
-            // Approve the offchain transfer from the account that's withdrawing back to the liquidity provider
-            IExchangeV3(fastWithdrawal.exchange).approveOffchainTransfer(
-                fastWithdrawal.from,
-                liquidityProvider,
-                fastWithdrawal.token,
-                fastWithdrawal.amount,
-                fastWithdrawal.feeToken,
-                fastWithdrawal.fee,
-                fastWithdrawal.validUntil,
-                fastWithdrawal.nonce
-            );
+        // Check the signature
+        require(hash.verifySignature(fwa.provider, fwa.signature ), "INVALID_SIGNATURE");
+        require(checkValidUntil(fwa.validUntil), "FASTWITHDRAWAL_APPROVAL_EXPIRED");
+
+        IExchangeV3 exchange = IExchangeV3(fwa.exchange);
+
+        address recipient = exchange.getWithdrawalRecipient(
+            fwa.from,
+            fwa.to,
+            fwa.token,
+            fwa.amount,
+            fwa.storageID
+        );
```

Comment:
Much more efficient to do try/catch on `setWithdrawalRecipient` instead. Saves several calls and storage reads.

Reply:
Ok, fixed.